network/write_buffer.go

```diff
@@ -0,0 +1,167 @@
+package network
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/golang/snappy"
+	"github.com/grafana/walqueue/types"
+	"go.uber.org/atomic"
+)
+
+// writeBuffer handles buffering the data, keeping track if there is a write request already running and kicking off the
+// write request as needed. All methods need to be called in a thread safe manner.
+type writeBuffer[T types.Datum] struct {
+	items []T
+	// writeAvailable keeps track if there is a write request going out.
```

Comment:
Naming of this feels off - should this be writeInProgress? Or is it meant to scale in the future so you can have N write requests out, and the buffer is only "not available" if you have N in progress? Maybe bufferFull?

Reply:
Sounds reasonable for writeInProgress, flipping the checks for it.